File: NueralNetworkLangauge/bindings/csharp/NeuroLang.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace NeuroLang
{
    /// <summary>
    /// NeuroLang C#/.NET SDK — Neural networks in plain English.
    ///
    /// Usage:
    /// <code>
    /// var nl = new NeuroLangClient("http://localhost:3000");
    ///
    /// // One-liner
    /// var result = await nl.RunAsync("Predict species from iris");
    /// Console.WriteLine($"Accuracy: {result.Accuracy}");
    ///
    /// // Builder pattern
    /// var result = await nl.Predict("species")
    ///     .From("iris")
    ///     .WithFeatures("petal_length", "petal_width")
    ///     .Epochs(30)
    ///     .TrainAsync();
    ///
    /// // Generate code for other frameworks
    /// string pyCode = await nl.PyTorchAsync("Predict species from iris");
    ///
    /// // With your own data
    /// var data = new[] {
    ///     new Dictionary&lt;string, double&gt; { ["age"] = 25, ["income"] = 50000, ["churned"] = 0 },
    ///     new Dictionary&lt;string, double&gt; { ["age"] = 45, ["income"] = 120000, ["churned"] = 1 },
    /// };
    /// var result = await nl.Predict("churned")
    ///     .WithFeatures("age", "income")
    ///     .WithData(data)
    ///     .TrainAsync();
    /// </code>
    /// </summary>
    public class NeuroLangClient
    {
        private readonly HttpClient _http;
        private readonly string _baseUrl;

        public NeuroLangClient(string baseUrl = "http://localhost:3000")
        {
            _baseUrl = baseUrl.TrimEnd('/');
            _http = new HttpClient { Timeout = TimeSpan.FromMinutes(5) };
        }

        public async Task<TrainResult> RunAsync(string source)
        {
            var payload = new { source };
            return await PostAsync<TrainResult>("/run", payload);
        }

        public async Task<CompileResult> CompileAsync(string source, string target = "tensorflow")
        {
            var payload = new { source, target };
            return await PostAsync<CompileResult>("/compile", payload);
        }

        public async Task<string> PyTorchAsync(string source)
            => (await CompileAsync(source, "pytorch")).Code;

        public async Task<string> KerasAsync(string source)
            => (await CompileAsync(source, "keras")).Code;

        public async Task<string> JaxAsync(string source)
            => (await CompileAsync(source, "jax")).Code;

        public ModelBuilder Predict(string target) => new ModelBuilder(this, target);

        private async Task<T> PostAsync<T>(string path, object payload)
        {
            var json = JsonSerializer.Serialize(payload, new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase
            });
            var content = new StringContent(json, Encoding.UTF8, "application/json");
            var response = await _http.PostAsync($"{_baseUrl}{path}", content);

            var body = await response.Content.ReadAsStringAsync();
            if (!response.IsSuccessStatusCode)
                throw new Exception($"NeuroLang API error ({(int)response.StatusCode}): {body}");

            return JsonSerializer.Deserialize<T>(body, new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase
            })!;
        }

        // ── Result Types ──────────────────────────────────

        public class TrainResult
        {
            public double? Accuracy { get; set; }
            public double? Mse { get; set; }
            public double FinalLoss { get; set; }
            public int Epochs { get; set; }
            public string Code { get; set; } = "";
            public List<Dictionary<string, object>>? Predictions { get; set; }

            public override string ToString()
            {
                if (Accuracy.HasValue) return $"TrainResult {{ Accuracy={Accuracy:F4}, Epochs={Epochs} }}";
                if (Mse.HasValue) return $"TrainResult {{ MSE={Mse:F6}, Epochs={Epochs} }}";
                return $"TrainResult {{ Loss={FinalLoss:F4}, Epochs={Epochs} }}";
            }
        }

        public class CompileResult
        {
            public string Code { get; set; } = "";
            public string Target { get; set; } = "tensorflow";
            public JsonElement Ir { get; set; }
        }

        // ── Builder ───────────────────────────────────────

        public class ModelBuilder
        {
            private readonly NeuroLangClient _client;
            private string _target;
            private string? _dataset;
            private List<string> _features = new();
            private int? _epochs;
            private int? _batchSize;
            private double? _learningRate;
            private string? _learnMode;
            private IEnumerable<Dictionary<string, double>>? _data;

            internal ModelBuilder(NeuroLangClient client, string target)
            {
                _client = client;
                _target = target;
            }

            public ModelBuilder From(string dataset) { _dataset = dataset; return this; }
            public ModelBuilder WithFeatures(params string[] features) { _features.AddRange(features); return this; }
            public ModelBuilder Epochs(int n) { _epochs = n; return this; }
            public ModelBuilder BatchSize(int n) { _batchSize = n; return this; }
            public ModelBuilder LearningRate(double lr) { _learningRate = lr; return this; }
            public ModelBuilder Deep() { _learnMode = "deep"; return this; }
            public ModelBuilder Linear() { _learnMode = "linear"; return this; }
            public ModelBuilder Auto() { _learnMode = "auto"; return this; }
            public ModelBuilder WithData(IEnumerable<Dictionary<string, double>> data) { _data = data; return this; }

            public string BuildSource()
            {
                var lines = new List<string> { $"predict {_target}" };
                if (_features.Count > 0) lines.Add($"inputs {string.Join(" ", _features)}");
                if (_dataset != null) lines.Add($"dataset {_dataset}");
                if (_epochs.HasValue) lines.Add($"epochs {_epochs}");
                if (_batchSize.HasValue) lines.Add($"batch_size {_batchSize}");
                if (_learningRate.HasValue) lines.Add($"learning_rate {_learningRate}");
                if (_learnMode != null) lines.Add($"learn {_learnMode}");
                return string.Join("\n", lines);
            }

            public async Task<TrainResult> TrainAsync()
            {
                var payload = new Dictionary<string, object> { ["source"] = BuildSource() };
                if (_data != null) payload["data"] = _data;
                return await _client.PostAsync<TrainResult>("/run", payload);
            }

            public async Task<CompileResult> CompileAsync()
            {
                return await _client.CompileAsync(BuildSource());
            }
        }
    }
}

File: NueralNetworkLangauge/bindings/go/neurolang.go
// Package neurolang provides a Go client for the NeuroLang neural network language.
//
// Write neural networks in plain English from Go.
//
// Usage:
//
//	client := neurolang.New("http://localhost:3000")
//
//	// One-liner
//	result, err := client.Run("Predict species from iris")
//	fmt.Printf("Accuracy: %.4f\n", *result.Accuracy)
//
//	// Builder pattern
//	result, err := client.Predict("species").
//	    From("iris").
//	    WithFeatures("petal_length", "petal_width").
//	    Epochs(30).
//	    Train()
//
//	// With your own data
//	data := []map[string]float64{
//	    {"age": 25, "income": 50000, "churned": 0},
//	    {"age": 45, "income": 120000, "churned": 1},
//	}
//	result, err := client.Predict("churned").
//	    WithFeatures("age", "income").
//	    WithData(data).
//	    Train()
//
//	// Generate code
//	code, err := client.PyTorch("Predict species from iris")
//	fmt.Println(code)
package neurolang

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type Client struct {
	BaseURL    string
	HTTPClient *http.Client
}

func New(baseURL string) *Client {
	return &Client{
		BaseURL: baseURL,
		HTTPClient: &http.Client{
			Timeout: 5 * time.Minute,
		},
	}
}

// TrainResult holds the result of compiling and training a NeuroLang program.
type TrainResult struct {
	Accuracy    *float64                 `json:"accuracy"`
	MSE         *float64                 `json:"mse"`
	FinalLoss   float64                  `json:"finalLoss"`
	Epochs      int                      `json:"epochs"`
	Code        string                   `json:"code"`
	Predictions []map[string]interface{} `json:"predictions"`
}

// CompileResult holds the result of compiling a NeuroLang program.
type CompileResult struct {
	Code   string                 `json:"code"`
	Target string                 `json:"target"`
	IR     map[string]interface{} `json:"ir"`
}

// Run compiles and trains a NeuroLang program.
func (c *Client) Run(source string) (*TrainResult, error) {
	payload := map[string]interface{}{"source": source}
	var result TrainResult
	if err := c.post("/run", payload, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// Compile compiles a NeuroLang program without training.
func (c *Client) Compile(source string, target ...string) (*CompileResult, error) {
	t := "tensorflow"
	if len(target) > 0 {
		t = target[0]
	}
	payload := map[string]interface{}{"source": source, "target": t}
	var result CompileResult
	if err := c.post("/compile", payload, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// PyTorch generates PyTorch code from NeuroLang source.
func (c *Client) PyTorch(source string) (string, error) {
	result, err := c.Compile(source, "pytorch")
	if err != nil {
		return "", err
	}
	return result.Code, nil
}

// Keras generates Keras code from NeuroLang source.
func (c *Client) Keras(source string) (string, error) {
	result, err := c.Compile(source, "keras")
	if err != nil {
		return "", err
	}
	return result.Code, nil
}

// JAX generates JAX/Flax code from NeuroLang source.
func (c *Client) JAX(source string) (string, error) {
	result, err := c.Compile(source, "jax")
	if err != nil {
		return "", err
	}
	return result.Code, nil
}

// Predict starts building a model with a builder pattern.
func (c *Client) Predict(target string) *ModelBuilder {
	return &ModelBuilder{client: c, target: target}
}

// ModelBuilder provides a fluent API for constructing NeuroLang programs.
type ModelBuilder struct {
	client       *Client
	target       string
	dataset      string
	features     []string
	epochs       int
	batchSize    int
	learningRate float64
	learnMode    string
	inlineData   []map[string]float64
}

func (b *ModelBuilder) From(dataset string) *ModelBuilder {
	b.dataset = dataset
	return b
}

func (b *ModelBuilder) WithFeatures(features ...string) *ModelBuilder {
	b.features = append(b.features, features...)
	return b
}

func (b *ModelBuilder) Epochs(n int) *ModelBuilder {
	b.epochs = n
	return b
}

func (b *ModelBuilder) BatchSize(n int) *ModelBuilder {
	b.batchSize = n
	return b
}

func (b *ModelBuilder) LearningRate(lr float64) *ModelBuilder {
	b.learningRate = lr
	return b
}

func (b *ModelBuilder) Deep() *ModelBuilder {
	b.learnMode = "deep"
	return b
}

func (b *ModelBuilder) Linear() *ModelBuilder {
	b.learnMode = "linear"
	return b
}

func (b *ModelBuilder) Auto() *ModelBuilder {
	b.learnMode = "auto"
	return b
}

func (b *ModelBuilder) WithData(data []map[string]float64) *ModelBuilder {
	b.inlineData = data
	return b
}

func (b *ModelBuilder) BuildSource() string {
	var lines []string
	if b.target != "" {
		lines = append(lines, "predict "+b.target)
	}
	if len(b.features) > 0 {
		lines = append(lines, "inputs "+strings.Join(b.features, " "))
	}
	if b.dataset != "" {
		lines = append(lines, "dataset "+b.dataset)
	}
	if b.epochs > 0 {
		lines = append(lines, fmt.Sprintf("epochs %d", b.epochs))
	}
	if b.batchSize > 0 {
		lines = append(lines, fmt.Sprintf("batch_size %d", b.batchSize))
	}
	if b.learningRate > 0 {
		lines = append(lines, fmt.Sprintf("learning_rate %g", b.learningRate))
	}
	if b.learnMode != "" {
		lines = append(lines, "learn "+b.learnMode)
	}
	return strings.Join(lines, "\n")
}

func (b *ModelBuilder) Train() (*TrainResult, error) {
	source := b.BuildSource()
	payload := map[string]interface{}{"source": source}
	if b.inlineData != nil {
		payload["data"] = b.inlineData
	}
	var result TrainResult
	if err := b.client.post("/run", payload, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

func (b *ModelBuilder) Compile() (*CompileResult, error) {
	return b.client.Compile(b.BuildSource())
}

func (c *Client) post(path string, payload interface{}, out interface{}) error {
	body, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("neurolang: marshal error: %w", err)
	}

	resp, err := c.HTTPClient.Post(c.BaseURL+path, "application/json", bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("neurolang: connection error (is the server running at %s?): %w", c.BaseURL, err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("neurolang: read error: %w", err)
	}

	if resp.StatusCode >= 400 {
		return fmt.Errorf("neurolang: API error %d: %s", resp.StatusCode, string(respBody))
	}

	return json.Unmarshal(respBody, out)
}
